test_util/helpers.py

```diff
@@ -23,81 +24,139 @@ def path_join(p1, p2):
 
-        # allow kwarg to override verification so client can be used generically
-        if self.ca_cert_path and 'verify' not in kwargs:
-            kwargs['verify'] = self.ca_cert_path
+    def get_client(self, scheme=None, host=None, path=None, query=None, fragment=None, port=None):
+        """Takes the same arguments as Url.get_url
+        """
+        clone = copy.deepcopy(self)
+        clone.url = self.url.get_url(scheme=scheme, host=host, path=path, query=query, fragment=fragment, port=port)
+        return clone
 
-        if self.api_base:
-            path = path_join(self.api_base, path)
+    def api_request(self, method, path_ex, scheme=None, host=None, query=None,
+                    fragment=None, path=None, port=None, node=None, **kwargs):
+        """ Direct wrapper for session.request. Returns request.Response
+        Args:
+            method: the HTTP request method to be used
+            path_ex: the extension to the path that is set as the default Url
 
+        Keyword Args:
+            All the named keyword args except node are identical to Url.get_url
+            node: can only be used if a get_node_port is set.
+            **kwargs: anything that can be passed to requests.request
+        """
         if node is not None:
-            assert host_url is None, 'Cannot set both node ({}) and host_url ({})'.format(node, host_url)
-            assert self._get_node_url is not None, 'get_node_url function must be supplied'
-            host_url = self._get_node_url(node, port=port)
-        else:
-            host_url = host_url if host_url else self.default_host_url
-        request_url = path_join(host_url, path)
-        headers.update(kwargs.pop('headers', {}))
-        logging.info('Request method {}: {}'.format(method, request_url))
-        logging.debug('Reqeust kwargs: {}'.format(kwargs))
-        logging.debug('Request headers: {}'.format(headers))
-        return requests.request(method, request_url, headers=headers, **kwargs)
+            assert port is None, 'node is intended to retrieve port; cannot set both simultaneously'
+            assert host is None, 'node is intended to retrieve host; cannot set both simultaneously'
+            assert self._get_node_port is not None, 'get_node_netloc must be specified for this ApiClient!'
+            port = self._get_node_port(node)
+            # do not explicitly declare default ports
+            host = node
+            if (port == 80 and self.url.scheme == 'http') or (port == 443 and self.url.scheme == 'https'):
+                port = None
 
-    def get_client(self, path, default_headers=None):
-        new_client = copy.deepcopy(self)
-        if default_headers is not None:
-            new_client.default_headers.update(default_headers)
-        new_client.api_base = path_join(self.api_base, path) if self.api_base is not None else path
-        return new_client
+        final_path = path_join(path if path else self.url.path, path_ex)
```

Comment:
> the request wrapper accepts the same paraeters as Url 

Maybe changing _that_ is the key (maybe `api_request()` should not receive `path` argument, but _just_ `path_extension`).

Guarantees only help if they are _simple_ (easy to understand). If an ApiClient object _guarantees_ to keep a certain URL "path prefix" constant, then that's simple. This has semantics that allows for writing solid tests. Then I can keep in mind "a client object actually _represents_ certain URL settings and client state, such as authentication.

If the URL path prefix later on extensible with a suffix, that's also easy (and actually _required_ to make ApiClient usable).

However, if we can later on override this URL prefix in `api_request()`, that's not simple anymore. Then, there are actually two methods for achieving the same thing: set `path` _or_ `path_extension`. Then, the ApiClient object actually represents a little less state that I can rely on. And I wonder why I should inject URL settings in the first place, upon ApiClient construction.

Pretty sure that needs to be consolidated, so that we get closer to the principle of "_one_ way to achieve a given thing."